### DISCUSSION TITLE
Speculative sampling?

### DISCUSSION THREAD

**Discussion by ghost** [ggml-org/llama.cpp#3863] [General]
Claims 2-2.5x speedup by getting multiple tokens per transformer call, could help a lot with large models living mostly in RAM vs VRAM. https://arxiv.org/abs/2302.01318.  Seems textsynth already has this feature.

**Comment by KerfuffleV2** [comment]
Look in `examples/speculative`

**Reply by ghost** [reply]
thank  you

**Comment by ghost** [comment]
So I guess to actually use it there would need to be a draft model in addition to the full transformer and both would have to be loaded into mem in unison, then it would be necessary to have a server parm which selected the number of spec samples if I understood the example correctly.  Don't completely understand where the draft model is supposed to come from though.

**Reply by KerfuffleV2** [reply]
Basically you choose a smaller model in the same family. Since it's in the same family, it will have been trained on similar tokens and (presumably) will make similar choices to the larger model.

There's also self-speculative sampling where you just don't run some parts of the main model to create the "draft" one: #3435 

I also have this pull #3565 (not really worth trying out though). It's a proof of concept for self-speculation that allows skipping layers, however I haven't been able to get effective results. In fact, it's worse than just not using speculation at all. (Note, this isn't really the full self-speculative sampling described in the paper, just an example of skipping layers in a similar way so it doesn't do anything like disprove self-speculation is useful.)

**Reply by ghost** [reply]
Thanks.  I just dug through the cornell paper and came away with the same understanding that you just described, they seem to recommend also just using a smaller version of the same model, though I guess in theory it would be possible to draft with a low quant mistral and see if some of the knowledge of falcon 180B loaded  full into main mem can be extracted faster, 2.5x average is a lot of speedup.  This approach seems quite interesting to help the memory bw bottleneck issue.

**Reply by KerfuffleV2** [reply]
> though I guess in theory it would be possible to draft with a low quant mistral and see if some of the knowledge of falcon 180B l

Maybe, just keep in mind that whatever the draft model suggests has to be fed to the full model. This can be done as a batch, but still if you mispredict too much stuff even with a very fast draft model it's likely to make the whole speculation stuff not worth it.

It also kind of seems like you need a draft model that works well enough to frequently draft decently long runs of tokens, rather than accurately drafting one token, feeding it to the main model, accurately drafting one token, etc. Just as an example, with my self-speculation stuff, even with 50% coming from the draft model and an 80% success success in predictions it still wasn't even enough to break even (that was testing with a 70B).

**Reply by ghost** [reply]
I suspect the secret sauce here is in creating a good draft model.  Sort of like downscaling an image becomes fuzzy but is much smaller but still has the same basic picture need a general scheme to downscale a model making it more fuzzy but still the same basic model (more than just quantizing it to low precision coefs).  Most likely the mistral + falcon is a bad idea but a downsampled falcon180 is what would needed (not falcon 7B, but falcon 180B downsampled to 7B size by some smart algorithm).  I guess that would be an entire new area of research needed to help make this idea real.  Decimating layers most likely was too coarse but its in line with the general idea of making a fuzzy smaller version of a bigger model.  Alternately some models might be coming out with drafts built in in the future as the paper suggested which would solve the problem also up front, which is most likely the best approach of all.

**Reply by KerfuffleV2** [reply]
> Alternately some models might be coming out with drafts built in in the future as the paper suggested which would solve the problem also up front, which is most likely the best approach of all.

Medusa sort of works like that: https://github.com/FasterDecoding/Medusa

**Reply by ghost** [reply]
Yeah I think that is the way.  When models are being creating they need to train a draft and the main model at same time off the same data as you alluded earlier.  Will be interesting to see if the mistral guys leverage this idea in their next release (maybe a 30B main and 7B draft combo).  My guess is decimating an existing model is going to be a very hard problem but creating draft and main at training time gets around the whole problem.
